src/h3lib/lib/algos.c

```diff
@@ -634,28 +634,126 @@ int H3_EXPORT(hexRing)(H3Index origin, int k, H3Index* out) {
 int H3_EXPORT(maxPolyfillSize)(const GeoPolygon* geoPolygon, int res) {
     // Get the bounding box for the GeoJSON-like struct
     BBox bbox;
-    bboxFromGeofence(&geoPolygon->geofence, &bbox);
-    int minK = bboxHexRadius(&bbox, res);
-
-    // The total number of hexagons to allocate can now be determined by
-    // the k-ring hex allocation helper function.
-    return H3_EXPORT(maxKringSize)(minK);
+    const Geofence geofence = geoPolygon->geofence;
+    bboxFromGeofence(&geofence, &bbox);
+    int numHexagons = bboxHexEstimate(&bbox, res);
+    // This algorithm assumes that the number of vertices is usually less than
+    // the number of hexagons, but when it's wrong, this will keep it from
+    // failing
+    int numVerts = geofence.numVerts;
+    for (int i = 0; i < geoPolygon->numHoles; i++) {
+        numVerts += geoPolygon->holes[i].numVerts;
+    }
+    if (numHexagons < numVerts) numHexagons = numVerts;
+    return numHexagons;
 }
 
 /**
  * polyfill takes a given GeoJSON-like data structure and preallocated,
  * zeroed memory, and fills it with the hexagons that are contained by
  * the GeoJSON-like data structure.
  *
- * The current implementation is very primitive and slow, but correct,
- * performing a point-in-poly operation on every hexagon in a k-ring defined
- * around the given geofence.
+ * This implementation traces the GeoJSON geofence(s) with hexagons, tests
+ * them and their neighbors to be contained by the geofence(s), and then any
+ * newly found hexagons are used to test again until no new hexagons are found.
```

Comment:
Nit: It might be worth noting here that the algorithm treats all polygons as cartesian, rather than spherical, since we're doubling down on that choice with the edge tracing.

Reply:
Will do.